example_client_test.go

```diff
@@ -0,0 +1,215 @@
+	queue := Queue{
+		logger: log.New(os.Stdout, "", log.LstdFlags),
+		name:   name,
+		done:   make(chan bool),
+	}
+	go queue.handleReconnect(addr)
+	return &queue
+}
+
+// handleReconnect will wait for a connection error on
+// notifyClose, and then continously attempt to reconnect.
+func (queue *Queue) handleReconnect(addr string) {
+	for {
+		queue.isConnected = false
+		log.Println("Attempting to connect")
+		for !queue.connect(addr) {
+			log.Println("Failed to connect. Retrying...")
+			time.Sleep(reconnectDelay)
+		}
+		select {
+		case <-queue.done:
+			return
+		case <-queue.notifyClose:
+		}
+	}
+}
+
+// connect will make a single attempt to connect to
+// RabbitMQ. It returns the success of the attempt.
+func (queue *Queue) connect(addr string) bool {
+	conn, err := amqp.Dial(addr)
+	if err != nil {
+		return false
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return false
+	}
+	ch.Confirm(false)
+	_, err = ch.QueueDeclare(
+		queue.name,
+		false, // Durable
+		false, // Delete when unused
+		false, // Exclusive
+		false, // No-wait
+		nil,   // Arguments
+	)
+	if err != nil {
+		return false
+	}
+	queue.changeConnection(conn, ch)
+	queue.isConnected = true
+	log.Println("Connected!")
+	return true
+}
+
+// changeConnection takes a new connection to the queue,
+// and updates the channel listeners to reflect this.
+func (queue *Queue) changeConnection(connection *amqp.Connection, channel *amqp.Channel) {
+	queue.connection = connection
+	queue.channel = channel
+	queue.notifyClose = make(chan *amqp.Error)
+	queue.notifyConfirm = make(chan amqp.Confirmation)
+	queue.channel.NotifyClose(queue.notifyClose)
+	queue.channel.NotifyPublish(queue.notifyConfirm)
+}
+
+// Push will push data onto the queue, and wait for a confirm.
+// If no confirms are recieved until within the resendTimeout,
+// it continuously resends messages until a confirm is recieved.
+// This will block until the server sends a confirm. Errors are
+// only returned if the push action itself fails, see UnsafePush.
+func (queue *Queue) Push(data []byte) error {
+	if !queue.isConnected {
+		return errors.New("failed to push push: not connected")
+	}
+	for {
+		err := queue.UnsafePush(data)
+		if err != nil {
+			queue.logger.Println("Push failed. Retrying...")
+			continue
+		}
+		select {
+		case confirm := <-queue.notifyConfirm:
+			if confirm.Ack {
+				queue.logger.Println("Push confirmed!")
+				return nil
+			}
+		case <-time.After(resendDelay):
+		}
+		queue.logger.Println("Push didn't confirm. Retrying...")
+	}
+}
+
+// UnsafePush will push to the queue without checking for
+// confirmation. It returns an error if it fails to connect.
+// No guarantees are provided for whether the server will
+// recieve the message.
+func (queue *Queue) UnsafePush(data []byte) error {
+	if !queue.isConnected {
+		return errNotConnected
+	}
+	return queue.channel.Publish(
+		"",         // Exchange
+		queue.name, // Routing key
+		false,      // Mandatory
```

Comment:
You may consider setting mandatory to be `true` in this example as it will ensure that messages can be routed. The fact that RabbitMQ drops non-routable messages is a very common source of confusion for new users.

Reply:
Up to you. Or, a comment could be made in the code.